github.js

```diff
@@ -338,56 +317,124 @@ GithubLocation.prototype = {
   // { versions: { versionhash } }
   // { notfound: true }
   lookup: function(repo) {
-    var execOpt = this.execOpt;
+    var self = this;
     var remoteString = this.remoteString;
-    return new Promise(function(resolve, reject) {
-      execGit('ls-remote ' + remoteString.replace(/(['"()])/g, '\\\$1') + repo + '.git refs/tags/* refs/heads/*', execOpt, function(err, stdout, stderr) {
-        if (err) {
-          if (err.toString().indexOf('not found') == -1) {
-            var error = new Error(stderr);
-            error.hideStack = true;
-            error.retriable = true;
-            reject(error);
-          }
+
+    return Promise.resolve()
+    .then(function() {
+      if (self.auth && self.auth.token) {
+        // use API to get branches/tags
+        return Promise.all(['tags', 'heads'].map(function(type) {
+          return asp(request)(extend({
+            uri: self.apiRemoteString + 'repos/' + repo + '/git/refs/' + type + self.authSuffix,
+            headers: {
+              'Accept': 'application/vnd.github.v3.raw'
+            }
+          }, self.defaultRequestOptions));
+        })).then(function(responses) {
+          var tagRes = responses[0];
+          var headRes = responses[1];
+
+          var refs = [];
+
+          // there should always be heads
+          if (headRes.statusCode != 200)
+            throw { statusCode: headRes.statusCode, headers: headRes.headers, api: true };
           else
-            resolve({ notfound: true });
-        }
+            refs = refs.concat(JSON.parse(headRes.body));
+
+          // tag response can be 404, i.e. no tags
+          if (tagRes.statusCode == 200)
+            refs = refs.concat(JSON.parse(tagRes.body));
+          else if (tagRes.statusCode != 404)
+            throw { statusCode: tagRes.statusCode, headers: tagRes.headers, api: true };
 
-        versions = {};
-        var refs = stdout.split('\n');
-        for (var i = 0; i < refs.length; i++) {
-          if (!refs[i])
-            continue;
+          return refs.map(function(obj) {
+            return { sha: obj.object.sha, name: obj.ref };
+          });
+        });
+      } else {
+        // fallback to git-based approach
+        return false;
+      }
+    })
+    .catch(function(e) {
+      if (e.headers && e.headers['x-ratelimit-remaining'] == '0') {
+        if (!apiWarned) {
+          ui.log('API ratelimit reached, falling back to slower git protocol');
```

Comment:
I was initially doing one request to get all references - tags/heads and pull requests. This proved very slow on large repos with lots of PRs.

at @tamird's suggestion I then did two requests, one to fetch tags, one for heads. This is now the same speed or faster than git, and can be cached (but isn't cached currently in my implementation)